man/man3/get_vaddr_for_device.3
\" -*- nroff -*-

\" get virtual address for board by name
\"
\" This program is free software; you can redistribute it and/or modify
\" it under the terms of the GNU General Public License as published by
\" the Free Software Foundation; either version 2 of the License, or (at
\" your option) any later version.
\"
\" This program is distributed in the hope that it will be useful, but
\" WITHOUT ANY WARRANTY; without even the implied warranty of
\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
\" General Public License for more details.
\"
\" You should have received a copy of the GNU General Public License
\" along with this program. If not, see <http://www.gnu.org/licenses/>.
\"
\" Author: Dean W. Anneser
\" Company: RTLinux Solutions LLC for Highland Technology, Inc.
\" Date: 30 Apr 2021

.TH get_vaddr_for_device 3 "get virtual address for board by name"

.SH NAME
\f4get_vaddr_for_device\f1 - get virtual address for board by name

.SH SYNOPSIS
.nf
\f4void *get_vaddr_for_device(char *\f2conf_file, \f4char *\f2board_spec\f4)\f1
.fi

.SH DESCRIPTION
The \f4get_vaddr_for_device()\f1 routine returns a virtual address for a VME board specified with \f2board_spec\f1.  The
\f2board_spec\f1 is of the format: "\f2device_name[\f1repeater_num\f2][\f1board_num\f2]\f1".  For example, the third v450
board (starting from 0), in chassis/repeater/controller number 0 would be expressed as: "\f2v450[\f10\f2][\f12\f2]\f1".

This routine is used in conjuction with the \f4v120_config\f1 utility.  This utility reads a V120/VME configuration, configures
user specified A16/A24/A32 address spaces, probes for VME devices listed, and writes this to an output configuration file.  This 
filename is specified as \f2conf_file\f1 in the call to this routine.  In the output \f2conf_file\f1, the V120 devices are
listed as online or offline, and for each address space, have a start and end mapping register specified.  The VME devices are
listed as online or offline, and identify which V120 repeater and address space they are mapped through. 

The device name is one of the Highland Technology device names listed in \f2vme_search_specs_stor.h\f1 (structure definition is
in \f2vme_search_specs_sdef.h\f1).  Note, this file could be expanded by the user to include other VME devices. 

.SH DEVICE TYPES
.SS REPEATER
The only supported repeater type presently, is the Highland Technology V120 "v120".
.SS VME
The following is the list of Highland Technology VME boards presently supported by this interface:

.nf
.in +5
\f2v210\f1 -- VME Relay Module
\f2v220\f1 -- VME 4-20mA Analog Control I/O Module
\f2v230\f1 -- VME Analog Input Module
\f2v250\f1 -- VME Digital I/O Module
\f2v340\f1 -- VME Waveform Generator Module
\f2v344\f1 -- VME Waveform Generator
\f2v346\f1 -- VME Waveform Generator
\f2v350\f1 -- VME Analog/Digital Function Generator
\f2v360\f1 -- VME Tachometer Module
\f2v365\f1 -- VME Tachometer / Overspeed Module
\f2v370\f1 -- VME Waveform Generator
\f2v375\f1 -- VME Waveform Generator
\f2v380\f1 -- VME Strain Gauge / Weighing Module
\f2v385\f1 -- VME Strain Gauge / Load Cell Module
\f2v410\f1 -- VME RTD / Resistance Input Module
\f2v420\f1 -- VME Isolated Resistance Simulator
\f2v450\f1 -- VME Analog Input Module
\f2v460\f1 -- VME Analog Scanner Module
\f2v470\f1 -- VME Analog Output and Thermocouple Simulator Module
\f2v490\f1 -- VME Multi-range Digitizer
\f2v545\f1 -- VME Synchro/Resolver/LVDT/RVDT Simulation/Acquisition Module
.in -5
.fi

.SH EXAMPLES
For the following output config file:

.nf
.in +5
\f4repeater \f2v120[\f10\f2] \f4-a16 \f20xffff0000 \f4-s16 \f20x10000 \f4-sp16 \f2260 \f4-ep16 \f2261 \f4-a24 \f20xff000000 \f4-s24 \f20x100000 \f4-sp24 \f2196 \f4-ep24 \f2259 \f4-a32 \f20xe0000000 \f4-s32 \f20x200000 \f4-sp32 \f268 \f4-ep32 \f2195 \f4-o \f2true\f1
\f4repeater \f2v120[\f11\f2] \f4-a16 \f20xffff8000 \f4-s16 \f20x8000 \f4-a24 \f20xff100000 \f4-s24 \f20x100000 \f4-a32 \f20xe0200000 \f4-s32 \f20x200000 \f4-o \f2false\f1
\f4vmedevice \f2v545[\f10\f2][\f10\f2] \f4-base \f20xffff0000 \f4-a \f216 \f4-o \f2true         \f1# VME Synchro/Resolver/LVDT/RVDT Simulation/Acquisition Module
\f4vmedevice \f2v470[\f10\f2][\f10\f2] \f4-base \f20xffff0800 \f4-a \f216 \f4-o \f2true         \f1# VME Analog Output and Thermocouple Simulator Module
\f4vmedevice \f2v450[\f10\f2][\f10\f2] \f4-base \f20xffff0a00 \f4-a \f216 \f4-o \f2true         \f1# VME Analog Input Module
\f4vmedevice \f2v545[\f10\f2][\f11\f2] \f4-base \f20xffff1000 \f4-a \f216 \f4-o \f2false        \f1# VME Relay Module
\f4vmedevice \f2v545[\f10\f2][\f12\f2] \f4-base \f20xffff1800 \f4-a \f216 \f4-o \f2false        \f1# VME Relay Module
.in -5
.fi

The call \f4get_vaddr_for_device(\f2"conf_file"\f4, \f2"v450"\f4,\f2 0\f4)\f1, will construct a "v450[0]" string to lookup in \f4conf_file\f1, and then get the associated "v120[0]" device and A16 address space, and use the starting A16 page of \f2260\f1 and ending page of \f2261\f1 to map to that address space, and return a virtual address to the base address of the VME board.

.SH ERRORS
none

.SH VENDOR MANUALS
.nf
http://www.highlandtechnology.com/categories/vme.shtml
.fi

.SH SEE ALSO
\f4v120_config(1)\f1

.SH CAVEATS
none

.SH RESTRICTIONS
Up to four V120 devices supported per system.

.SH AUTHORS
Dean W. Anneser


lib/get_vaddr_for_device/get_vaddr_for_device.c
// -*- c++ -*-

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or (at
// your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
// Author: Dean W. Anneser
// Company: RTLinux Solutions LLC for Highland Technology, Inc.
// Date: Sat Aug 07 17:01:17 2021

#include <stdlib.h>
#include <string.h>

// structure defs

#include <V120.h>

// prototypes

#include "parse_args_proto.h"
#include "parse_options_proto.h"
#include "read_line_proto.h"

#define MAX_ARG_V 64
#define MAX_CMD_LINE 1024
#define MAX_FN_LENGTH 256

// configure VME cards

#define VME_SET_VME_BASE_ADDR		0x0000000000000001UL
#define VME_SET_REPEATER_NUM		0x0000000000000002UL
#define VME_SET_ADDR_SPACE		0x0000000000000004UL
#define VME_SET_ONLINE_FLAG		0x0000000000000008UL

#define REPEATER_SET_A16_ADDR		0x0000000000000001UL
#define REPEATER_SET_A16_SIZE		0x0000000000000002UL
#define REPEATER_SET_A16_STARTPAGE	0x0000000000000004UL
#define REPEATER_SET_A16_ENDPAGE	0x0000000000000008UL
#define REPEATER_SET_A24_ADDR		0x0000000000000010UL
#define REPEATER_SET_A24_SIZE		0x0000000000000020UL
#define REPEATER_SET_A24_STARTPAGE	0x0000000000000040UL
#define REPEATER_SET_A24_ENDPAGE	0x0000000000000080UL
#define REPEATER_SET_A32_ADDR		0x0000000000000100UL
#define REPEATER_SET_A32_SIZE		0x0000000000000200UL
#define REPEATER_SET_A32_STARTPAGE	0x0000000000000400UL
#define REPEATER_SET_A32_ENDPAGE	0x0000000000000800UL
#define REPEATER_SET_ONLINE_FLAG	0x0000000000001000UL

static void load_repeater_board_num(char *, int *, int *);

void *get_vaddr_for_device(char *conf_file, char *board_spec)
{
  char dev_spec[32];
  FILE *fpi;
  int arg_c;
  char *arg_v[MAX_ARG_V];
  char cmdline[MAX_CMD_LINE] = {""};
  char parsed_cmdline[MAX_CMD_LINE] = {""};
  int __attribute__((unused)) num_chars;
  V120_HANDLE *v120_handle;
  void *a16_base_vaddr, *a24_base_vaddr, *a32_base_vaddr;
  static void *vme_base_vaddr;

  static unsigned int vme_base_addr;
  static int vme_addr_space;
  static int vme_online_flag;
  static unsigned long vme_chgflg;

  int repeater_num, board_num;

  static OPTION_TABLE vme_option_table[] = {
    // opt_str		arg_str		arg_value	data				type_mask	routine			chgflg_mask             
    {"-base",		"*",		0,		(void *)&vme_base_addr,		T_UINT,		load_constant,		VME_SET_VME_BASE_ADDR},
    {"-a",		"*",		0,		(void *)&vme_addr_space,	T_INT,		load_constant,		VME_SET_ADDR_SPACE},
    {"-o",		"true",		1,		(void *)&vme_online_flag,	T_INT,		load_enum,		VME_SET_ONLINE_FLAG},
    {"-o",		"false",	0,		(void *)&vme_online_flag,	T_INT,		load_enum,		VME_SET_ONLINE_FLAG},
    {"",		"",		0,		(void *)0,			0,              0,                      0}
  };

  static unsigned int a16_addr;
  static unsigned int a16_size;
  static unsigned int a16_startpage;
  static unsigned int a16_endpage;
  static unsigned int a24_addr;
  static unsigned int a24_size;
  static unsigned int a24_startpage;
  static unsigned int a24_endpage;
  static unsigned int a32_addr;
  static unsigned int a32_size;
  static unsigned int a32_startpage;
  static unsigned int a32_endpage;
  static unsigned long repeater_online_flag;
  static unsigned long repeater_chgflg;

  static OPTION_TABLE repeater_option_table[] = {
    // opt_str		arg_str		arg_value	data				type_mask	routine			chgflg_mask             
    {"-a16",		"*",		0,		(void *)&a16_addr,	        T_UINT,		load_constant,		REPEATER_SET_A16_ADDR},      
    {"-s16",		"*",		0,		(void *)&a16_size,	        T_UINT,		load_constant,		REPEATER_SET_A16_SIZE},      
    {"-S16",		"*",		0,		(void *)0,			0,		noop,			0},
    {"-sp16",		"*",		0,		(void *)&a16_startpage,	        T_UINT,		load_constant,		REPEATER_SET_A16_STARTPAGE}, 
    {"-ep16",		"*",		0,		(void *)&a16_endpage,	        T_UINT,		load_constant,		REPEATER_SET_A16_ENDPAGE},   
    {"-a24",		"*",		0,		(void *)&a24_addr,	        T_UINT,		load_constant,		REPEATER_SET_A24_ADDR},      
    {"-s24",		"*",		0,		(void *)&a24_size,	        T_UINT,		load_constant,		REPEATER_SET_A24_SIZE},      
    {"-S24",		"*",		0,		(void *)0,			0,		noop,			0},
    {"-sp24",		"*",		0,		(void *)&a24_startpage,	        T_UINT,		load_constant,		REPEATER_SET_A24_STARTPAGE}, 
    {"-ep24",		"*",		0,		(void *)&a24_endpage,	        T_UINT,		load_constant,		REPEATER_SET_A24_ENDPAGE},   
    {"-a32",		"*",		0,		(void *)&a32_addr,	        T_UINT,		load_constant,		REPEATER_SET_A32_ADDR},      
    {"-s32",		"*",		0,		(void *)&a32_size,	        T_UINT,		load_constant,		REPEATER_SET_A32_SIZE},      
    {"-S32",		"*",		0,		(void *)0,			0,		noop,			0},
    {"-sp32",		"*",		0,		(void *)&a32_startpage,	        T_UINT,		load_constant,		REPEATER_SET_A32_STARTPAGE}, 
    {"-ep32",		"*",		0,		(void *)&a32_endpage,	        T_UINT,		load_constant,		REPEATER_SET_A32_ENDPAGE},   
    {"-o",		"true",		1,		(void *)&repeater_online_flag,	T_INT,		load_enum,		REPEATER_SET_ONLINE_FLAG},   
    {"-o",		"false",	1,		(void *)&repeater_online_flag,	T_INT,		load_enum,		REPEATER_SET_ONLINE_FLAG},   
  };

  vme_base_vaddr = (void *)0;

  load_repeater_board_num(board_spec, &repeater_num, &board_num);

  // open input configuration file 

  if ((fpi = fopen(conf_file, "r")) == (FILE *)0)
    {
      printf("FATAL: cannot open input configuration file: %s\n", conf_file);
      return((void *)-1);
    }
    
  while (((num_chars = read_line(cmdline, MAX_CMD_LINE, fpi)) > 0) && (vme_base_vaddr == (void *)0))
    {
      arg_c = parse_args(cmdline, arg_v, MAX_ARG_V, parsed_cmdline, MAX_CMD_LINE);
  
      if ((strcmp(arg_v[0], "vmedevice") == 0) && (strcmp(arg_v[1], board_spec) == 0))
        {
          // parse command line options/args -- load vme_base_addr, vme_addr_space, vme_online_flag

          parse_options(arg_c - 2, &arg_v[2], vme_option_table, &vme_chgflg);

          // if (vme_online_flag == 1)
              
          sprintf(dev_spec, "v120[%d]", repeater_num);

          rewind(fpi);

          while (((num_chars = read_line(cmdline, MAX_CMD_LINE, fpi)) > 0) && (vme_base_vaddr == (void *)0))
            {
              arg_c = parse_args(cmdline, arg_v, MAX_ARG_V, parsed_cmdline, MAX_CMD_LINE);

              if ((strcmp(arg_v[0], "repeater") == 0) && (strcmp(arg_v[1], dev_spec) == 0))
                {
                  // parse command line options/args -- load aXX_addr, aXX_size, aXX_startpage, aXX_endpage, repeater_online_flag

                  parse_options(arg_c - 2, &arg_v[2], repeater_option_table, &repeater_chgflg);
                  
                  // if (repeater_online_flag == 1)

                  v120_handle = v120_open(repeater_num);                  

                  switch(vme_addr_space)
                    {
                    case 16:
                      a16_base_vaddr = v120_get_vme(v120_handle, a16_startpage, a16_endpage);
                      vme_base_vaddr = (void *)(a16_base_vaddr + vme_base_addr - a16_addr);
                      break;

                    case 24:
                      a24_base_vaddr = v120_get_vme(v120_handle, a24_startpage, a24_endpage);
                      vme_base_vaddr = (void *)(a24_base_vaddr + vme_base_addr - a24_addr);
                      break;

                    case 32:
                      a32_base_vaddr = v120_get_vme(v120_handle, a32_startpage, a32_endpage);
                      vme_base_vaddr = (void *)(a32_base_vaddr + vme_base_addr - a32_addr);
                      break;

                    default:
                      break;
                    }
                }                
            }
        }
    }

  // close inout configuration file

  if (fpi != (FILE *)0)
    fclose(fpi);

  return(vme_base_vaddr);
}


static void load_repeater_board_num(char *board_spec, int *pcrate_num, int *pboard_num)
{
  char *ps = strdup(board_spec);
  char *p, *q;
  char __attribute__((unused)) *pname;

  p = ps;
  pname = strtok_r(p, "[]", &q);
  p = q;
  *pcrate_num = atoi(strtok_r(p, "[]", &q));
  p = q;
  *pboard_num = atoi(strtok_r(p, "[]", &q));
  free(ps);
  return;
}


